File: crates/j-law-ruby/test/test_income_tax.rb
# frozen_string_literal: true

# 所得税法第89条に基づく所得税計算のテスト。
#
# 法的根拠: 所得税法 第89条第1項 / 復興財源確保法 第13条
# テストケースは tests/fixtures/income_tax.json から読み込む。
#
# 実行方法:
#   rake compile && ruby -Ilib test/test_income_tax.rb

require "minitest/autorun"
require "json"
require "j_law_core"

INCOME_TAX_FIXTURES = JSON.parse(
  File.read(File.join(__dir__, "../../../tests/fixtures/income_tax.json"))
)

# ─── データ駆動テスト ─────────────────────────────────────────────────────────

class TestIncomeTaxFixtures < Minitest::Test
  INCOME_TAX_FIXTURES["income_tax"].each do |c|
    define_method("test_#{c['id']}") do
      inp = c["input"]
      exp = c["expected"]

      r = JLawCore::IncomeTax.calc_income_tax(
        inp["taxable_income"], inp["year"], inp["month"], inp["day"],
        inp["apply_reconstruction_tax"]
      )

      assert_equal exp["base_tax"], r.base_tax, "#{c['id']}: base_tax"
      assert_equal exp["reconstruction_tax"], r.reconstruction_tax, "#{c['id']}: reconstruction_tax"
      assert_equal exp["total_tax"], r.total_tax, "#{c['id']}: total_tax"
      if exp["reconstruction_tax_applied"]
        assert r.reconstruction_tax_applied?, "#{c['id']}: reconstruction_tax_applied"
      else
        refute r.reconstruction_tax_applied?, "#{c['id']}: reconstruction_tax_applied"
      end
    end
  end
end

# ─── 言語固有テスト ──────────────────────────────────────────────────────────

class TestIncomeTaxLanguageSpecific < Minitest::Test
  def test_error_date_out_of_range
    assert_raises(RuntimeError) do
      JLawCore::IncomeTax.calc_income_tax(5_000_000, 2014, 12, 31, true)
    end
  end

  def test_breakdown_fields
    r = JLawCore::IncomeTax.calc_income_tax(5_000_000, 2024, 1, 1, true)
    refute_empty r.breakdown
    r.breakdown.each do |step|
      refute_empty step[:label]
      assert_operator step[:rate_denom], :>, 0
    end
  end

  def test_inspect
    r = JLawCore::IncomeTax.calc_income_tax(5_000_000, 2024, 1, 1, true)
    assert_match(/IncomeTaxResult/, r.inspect)
  end
end

File: crates/j-law-python/tests/test_income_tax.py
"""所得税法第89条に基づく所得税計算のテスト。

法的根拠: 所得税法 第89条第1項 / 復興財源確保法 第13条

テストケースは tests/fixtures/income_tax.json から読み込む。
"""

import json
import pathlib

import pytest

from j_law_python.income_tax import calc_income_tax

# ─── フィクスチャ読み込み ─────────────────────────────────────────────────────

_FIXTURE_PATH = pathlib.Path(__file__).resolve().parents[3] / "tests" / "fixtures" / "income_tax.json"
_FIXTURES = json.loads(_FIXTURE_PATH.read_text(encoding="utf-8"))


# ─── データ駆動テスト ─────────────────────────────────────────────────────────


@pytest.mark.parametrize("case", _FIXTURES["income_tax"], ids=lambda c: c["id"])
def test_income_tax(case):
    inp = case["input"]
    exp = case["expected"]

    r = calc_income_tax(
        inp["taxable_income"],
        inp["year"],
        inp["month"],
        inp["day"],
        apply_reconstruction_tax=inp["apply_reconstruction_tax"],
    )

    assert r.base_tax == exp["base_tax"], f"{case['id']}: base_tax"
    assert r.reconstruction_tax == exp["reconstruction_tax"], f"{case['id']}: reconstruction_tax"
    assert r.total_tax == exp["total_tax"], f"{case['id']}: total_tax"
    assert r.reconstruction_tax_applied is exp["reconstruction_tax_applied"], f"{case['id']}: reconstruction_tax_applied"


# ─── 言語固有テスト（JSON の外） ──────────────────────────────────────────────


class TestLanguageSpecific:
    """Python 固有の振る舞い検証。"""

    def test_error_date_out_of_range(self):
        with pytest.raises(ValueError):
            calc_income_tax(5_000_000, 2014, 12, 31)

    def test_breakdown_fields(self):
        r = calc_income_tax(5_000_000, 2024, 1, 1)
        assert len(r.breakdown) > 0
        for step in r.breakdown:
            assert step.label != ""
            assert step.rate_denom > 0

    def test_repr(self):
        r = calc_income_tax(5_000_000, 2024, 1, 1)
        assert "IncomeTaxResult" in repr(r)
